File: SlidingWindow/longest_ones.py
from typing import List


class LongestOnes:
    def longestOnes(self, nums: List[int], k: int) -> int:
        max_consecutive = 0
        curr_zeros = 0
        left = 0
        for right in range(len(nums)):
            curr_zeros += 1 if nums[right] == 0 else 0
            while curr_zeros > k:
                curr_zeros -= 1 if nums[left] == 0 else 0
                left += 1
            max_consecutive = max(right - left + 1, max_consecutive)

        return max_consecutive

File: SlidingWindow/longestOnes.js
const longestOnes = (nums, k) => {
  let left = 0;
  let maxLength = 0;
  let zeroCount = 0;
  for (let right = 0; right < nums.length; right++) {
    if (nums[right] === 0) zeroCount++;

    while (zeroCount > k) {
      if (nums[left] === 0) zeroCount--;
      left++;
    }
    maxLength = Math.max(right - left + 1, maxLength);
  }
  return maxLength;
};

// Time Complexity: O(n) where n is the length of the input array
// Space Complexity: O(1) since we are using a constant amount of space for the variables
